### PR TITLE
DHIS2 API integration and JSON functionality

### PR DESCRIPTION
**Pull request by rdziewietin**
Features added (since repo creation):

- File upload
- Form type dropdown
- Data set, organization unit, and period inputs
- Editable tables (with ICCM hardcoded)
- DHIS2 API integration for all except data elements
- JSON download

### CODE REVIEW

File: data_upload_DHIS2.py
Comment (ginic) [inline data_upload_DHIS2.py]: Unless MSF would like to use logins or user authorization down the line, this seems like a good opportunity to use the Streamlit secrets management options, https://docs.streamlit.io/deploy/streamlit-community-cloud/deploy-your-app/secrets-management. If you feel like changing that now, go for it, otherwise we will just keep this in mind for later improvements.

File: app.py
Comment (ginic) [inline app.py]: Thanks for adding docstrings! For python, docstring comments always go after the function/class signature. In this case, pleas just copy the docstring to the line after `def convert(dfs)`. You can see more info at https://realpython.com/documenting-python-code/ and https://peps.python.org/pep-0257/

File: app.py
Comment (ginic) [inline app.py]: Typo fix: `param df` should match the name of the argument in the function signature, so it should be `param dfs`

File: app.py
Comment (ginic) [inline app.py]: Whoohoo for docstrings!! Just move it to be after the function definition line (see my comment below).

File: app.py
Comment (ginic) [inline app.py]: I'm surprised we want to accept multiple files here, I thought most of the tally sheets were only one page, but now that I'm looking back at them, I'm realizing that isn't true. It could be a challenge for the system if users upload things in a different order than normal. No code changes needed for now, but example to add to our list for testing later.

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks, @rdziewietin! Overall this looks great. Nice work! I was able to run the app in Docker, simulate uploading and inputting a form, then download the JSON file. The JSON output looked good to me, but I'll defer to @anjus1313 as the expert there. 

Regarding the UI itself, I was wondering what the best way for the user to start over to upload a new form is? We might want to add a button for that later, something like "Upload new form". 

I also had a few minor comments in the code that I'd like to have you take a look at. 

The only change that I think is necessary before you merge is to fix the docstrings. Feel free to address the other comments only if you have time this sprint. We can always put them as tasks in the backlog to be handled later.

File: data_upload_DHIS2.py
Comment (YajieLi-Z) [inline data_upload_DHIS2.py]: This function performs a fuzzy match based on the search terms provided by the user. For example, I searche for 'general OPD' and receive 50 matching results, the function may return the first match from a list of similar results. 

To ensure the ID is unique and accurate, it is recommended that the user provides the full and exact name. So I'm thinking if the search terms are incomplete, the user is prompted to refine the search or select the correct item from a list. Does that make sense?

**Review by YajieLi-Z — COMMENTED** [review]
The getUID function performs a fuzzy match based on the search terms provided by the user. For example, I search for 'general OPD' and receive 50 matching results, the function may return the first match from a list of similar results.

To ensure the ID is unique and accurate, it is recommended that the user provides the full and exact name. So I'm thinking if the search terms are incomplete, the user is prompted to refine the search or select the correct item from a list. Does that make sense?

File: data_upload_DHIS2.py
Comment (rdziewietin) [reply data_upload_DHIS2.py]: Yeah, definitely! With the OCR, I'm thinking that the pipeline would be OCR recognizes text, the user confirms the text is correct before searching (in case it is very wrong), the API searches for the ID, then prompts the user with a dropdown if it is not an exact match. Is that what you were thinking?

Comment (YajieLi-Z) [reply data_upload_DHIS2.py]: Yeah, this is exactly what I'm thinking~

**Pull request merged**
